mountpoint-s3/src/inode.rs

```diff
@@ -1189,6 +1273,130 @@ mod tests {
         assert!(!client.contains_prefix(&prefix));
     }
 
+    // testing the case of concurrent deletion of parent before acquiring the lock on parent in create
+    #[test_case(""; "unprefixed")]
+    #[test_case("test_prefix/"; "prefixed")]
+    #[tokio::test]
+    async fn test_concurrent_create_rmdir(prefix: &str) {
+        let client_config = MockClientConfig {
+            bucket: "test_bucket".to_string(),
+            part_size: 1024 * 1024,
+        };
+        let client = Arc::new(MockClient::new(client_config));
+        let prefix = Prefix::new(prefix).expect("valid prefix");
+        let superblock = Superblock::new("test_bucket", &prefix);
+
+        // Create local directory
+        let dirname = "local_dir";
+        let LookedUp { inode, .. } = superblock
+            .create(&client, FUSE_ROOT_INODE, dirname.as_ref(), InodeKind::Directory)
+            .await
+            .unwrap();
+
+        // Marking deleted status as true to denote concurrent rmdir
+        {
+            let mut inode_state = inode.inner.sync.write().unwrap();
+
+            match &mut inode_state.kind_data {
+                InodeKindData::File {} => unreachable!("Created a directory"),
+                InodeKindData::Directory { deleted, .. } => *deleted = true,
+            }
+        }
+        // ending the lifetime of rwlock
+
+        // Create child directory
+        let child_dirname = "child_dir";
+        superblock
+            .create(&client, inode.ino(), child_dirname.as_ref(), InodeKind::Directory)
+            .await
+            .expect_err("Should not create sub-directory in a directory with deleted status");
+    }
+
+    #[test_case(""; "unprefixed")]
+    #[test_case("test_prefix/"; "prefixed")]
+    #[tokio::test]
+    async fn test_readdir_lookup_after_rmdir(prefix: &str) {
+        let client_config = MockClientConfig {
+            bucket: "test_bucket".to_string(),
+            part_size: 1024 * 1024,
+        };
+        let client = Arc::new(MockClient::new(client_config));
+        let prefix = Prefix::new(prefix).expect("valid prefix");
+        let superblock = Superblock::new("test_bucket", &prefix);
+
+        // Create local directory
+        let dirname = "local_dir";
+        let LookedUp { inode, .. } = superblock
+            .create(&client, FUSE_ROOT_INODE, dirname.as_ref(), InodeKind::Directory)
+            .await
+            .expect("Should be able to create directory");
+
+        superblock
+            .rmdir(&client, FUSE_ROOT_INODE, dirname.as_ref())
+            .await
+            .expect("rmdir on empty local directory should succeed");
+
+        superblock
+            .lookup(&client, FUSE_ROOT_INODE, dirname.as_ref())
+            .await
+            .expect_err("should not do lookup on removed directory");
```

Comment:
Shall we also test that `readdir` on the parent does not contain `dirname` anymore? (I think that's what was mentioned in other comments)